RocketChat/Rocket.Chat
# feat: node-gcm deprecation

## Proposed changes (including videos or screenshots)


Currently, we are using a library called [node-gcm](https://www.npmjs.com/package/node-gcm) to send native notifications via firebase. Unfortunately Firebase is deprecating a series of libraries related to notifications (see https://firebase.google.com/support/faq?hl=pt-br#fcm-23-deprecation), and, because of this, the library we used to use is not going to be updated, mostly in favor of using the [oficial firebase SDK](https://www.npmjs.com/package/firebase).
We discussed about the use of the oficial package, but the oficial package bundles up every firebase feature inside a huge library, even though we just use a comically small subset of the features (sending notifications).
Therefore, we decided on not using the Firebase SDK and just using HTTP requests to send native notifications.

## Backportability

This feature is tricky on this regard, because this API is going to stop working, so we need to add a feature flag (using a setting for it), because everyone who is using supported versions of Rocket.Chat need native notifications to work, so we add a flag called `Push_UseLegacy`, if it is set to true, we will use the old api (using `node-gcm`), if it is false, it is going to use our HTTP wrapper on the notification service.

## Issue(s)

#32124


## Steps to test or reproduce


First, it will be needed a way to connect a mobile device into the local server, you can choose whatever way you want, I am using a tunneling application called [ngrok](https://ngrok.com/).
So, first I will start the application, on port 3000 (default) and then I will start ngrok there, you can do it like this:
```sh
$ ngrok http http://localhost:3000
```

After running the command, you will have a terminal interface, and you will see the URL the application is forwarding your local webserver, copy that value or save it somewhere, it will be needed later!
(it is a weird URL, something like: https://9d8c-191-243-134-83.ngrok-free.app)

So now, to test it you will need a test build of the application, please @ me on open, I can send you the build I am using 😎 

Download the build and install it on your device.
Open the app, and it will ask you for the URL of the workspace, in this field, please, put the forwarded URL by ngrok.

Ok, so now, most of the config on the mobile side is done, now there are a few config's needed to enable the new notification provider:
1 - Go to settings>push
2 - Disable Push_UseLegacy
3 - Get the credentials JSON and paste it into Push_google_api_credentials

No we should be ready to go!
Just create a new user on your phone (don't have to, but it will be easier), and now, go on your workspace open in your computer, and ping the new user in the #general channel, you should get a notification on your phobe 

## Further comments


I've used https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages?hl=pt-br to type out the request type 😛 

CONN-73
